server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -409,34 +408,32 @@ public Integer readOptionalVInt() throws IOException {
     private static final ThreadLocal<byte[]> stringReadBuffer = ThreadLocal.withInitial(() -> new byte[1024]);
 
     // Thread-local buffer for smaller strings
-    private static final ThreadLocal<CharsRef> smallSpare = ThreadLocal.withInitial(() -> new CharsRef(SMALL_STRING_LIMIT));
+    private static final ThreadLocal<char[]> smallSpare = ThreadLocal.withInitial(() -> new char[SMALL_STRING_LIMIT]);
 
     // Larger buffer used for long strings that can't fit into the thread-local buffer
     // We don't use a CharsRefBuilder since we exactly know the size of the character array up front
     // this prevents calling grow for every character since we don't need this
-    private CharsRef largeSpare;
+    private char[] largeSpare;
 
-    public String readString() throws IOException {
+    public final String readString() throws IOException {
         final int charCount = readArraySize();
-        final CharsRef charsRef;
-        if (charCount > SMALL_STRING_LIMIT) {
-            if (largeSpare == null) {
-                largeSpare = new CharsRef(ArrayUtil.oversize(charCount, Character.BYTES));
-            } else if (largeSpare.chars.length < charCount) {
-                // we don't use ArrayUtils.grow since there is no need to copy the array
-                largeSpare.chars = new char[ArrayUtil.oversize(charCount, Character.BYTES)];
-            }
-            charsRef = largeSpare;
-        } else {
-            charsRef = smallSpare.get();
+        if (charCount == 0) {
+            return "";
         }
-        charsRef.length = charCount;
-        int charsOffset = 0;
+        final char[] charBuffer = charCount > SMALL_STRING_LIMIT ? ensureLargeSpare(charCount) : smallSpare.get();
+        readString(charCount, charBuffer);
+        return new String(charBuffer, 0, charCount);
+    }
+
+    protected void readString(int charCount, char[] charBuffer) throws IOException {
+        readStringSlow(0, charCount, charBuffer);
+    }
+
+    protected final void readStringSlow(int charsOffset, int charCount, char[] charBuffer) throws IOException {
```

Comment:
The args look like a `CharsRef` but they're not because the length is `charsLeft = charCount - charsOffset`. I suppose it might be marginally more efficient to pass in the original count rather than the remainder but could we make it clear that this API is a trap to folks who are used to working with `CharsRef`?

Reply:
Right this is rather confusing. I added some docs on this now.